keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1004 @@
+initially in the certificates/v1alpha1 API group.  The TrustAnchorSets object
+can be thought of as a specialized configmap tailored to the X.509 trust anchor
+use case.  Introducing a dedicated type allows us to attach different RBAC
+policies to TrustAnchorSets, which will typically be wide-open for reading, but
+locked-down for writing.
+
+Each TrustAnchorSets object is a container for multiple named trust anchor sets.
+Most use cases will simply using a single trust anchor set named default, but
+mutual TLS use cases involving multiple federating meshes will require the
+ability to express multiple, non-commingled trust anchor sets.
+
+Like CertificateSigningRequest, TrustAnchorSets are scoped to a particular
+signer using a `signerName` field, which will also be usable in RBAC
+ClusterRoleBindings.  It is possible to grant specific users the ability
+read and/or update all TrustAnchorSets that correspond to a specific signer.
+
+To enable consumption by workloads, the new `trustAnchorSets` Kubelet projected
+volume source allows projecting a trust anchor set from a particular signer into
+the container filesystem, with the contents of the projected files updating as
+the corresponding TrustAnchorSets are updated.
+
+### User Stories
+
+#### Trust Anchor Distribution for Private CAs
+
+In my cluster, I operate a certificates/v1 signer implementation called
+`example.com/server-tls`.  This is backed by a private CA hierarchy.  By some
+unspecified mechanism, I issue certificates from this CA hierarchy to a server
+workload (a pod named `server/server`).  My client workload (a pod named
+`client/client`) needs access to the trust anchors for the CA in order to
+connect to the server.  The client should be able to continually establish
+connections to the server without downtime, even as I rotate the trust anchors
+that back the CA.
+
+To do this, I update the controller that implements `example.com/server-tls` to
+create a TrustAnchorSets object like this one:
+```yaml
+apiVersion: v1alpha1
+kind: TrustAnchorSets
+metadata:
+  name: example-com-server-tls-abcdefg
+spec:
+  signerName: example.com/server-tls
+status:
+  sets:
+  - name: default
+    pemTrustAnchors:
+    - "<... PEM DATA ...>"
+```
+
+As the set of trust anchors for my CA changes due to rotations and revocations,
+my controller updates the status of the TrustAnchorSets object.
+
+I then update my client pod to add a `trustAnchorSets` volume:
+```diff
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: client
+  name: client
+spec:
+  containers:
+    name: main
+    image: my-image
+    volumeMounts:
++    - mountPath: /var/run/example-com-server-tls-trust-anchors
++      name: example-com-server-tls-trust-anchors
++      readOnly: true
+  volumes:
++  - name: example-com-server-tls-trust-anchors
++    projected:
++      sources:
++      - trustAnchorSets:
++          signerName: example.com/server-tls
```

Comment:
IIUC the kubelet is going to union all `trustAnchorSets` with this signer?  i.e. the client does not need to know the name of the `trustAnchorSet` object?

Reply:
Yes.  This is significantly simpler now that we use a fixed mapping between a signer name and a single TrustAnchorSets object.  I've also added an escape hatch that allows you to directly name a TrustAnchorSets object, in case you're doing something unrelated to signers.